File: 2022/cpp/day2.cpp
#include <algorithm>
#include <cstdint>
#include <fstream>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <ostream>
#include <string>
#include <vector>

int main() {
  std::fstream file;
  file.open("../inputs/input2", std::ios::in);
  if (file.is_open()) {

    std::map<std::string, std::uint32_t> play1;
    std::map<std::string, std::uint32_t> play2;
    play1 = {{"A X", 4}, {"A Y", 8}, {"A Z", 3}, {"B X", 1}, {"B Y", 5},
             {"B Z", 9}, {"C X", 7}, {"C Y", 2}, {"C Z", 6}};
    play2 = {{"A X", 3}, {"A Y", 4}, {"A Z", 8}, {"B X", 1}, {"B Y", 5},
             {"B Z", 9}, {"C X", 2}, {"C Y", 6}, {"C Z", 7}};

    std::string line;
    std::uint32_t part1 = 0;
    std::uint32_t part2 = 0;
    while (std::getline(file, line)) {
      part1 += play1[line];
      part2 += play2[line];
    }
    std::cout << "Part 1:" << part1 << std::endl;
    std::cout << "Part 2:" << part2 << std::endl;
  }
}

File: 2022/rs/src/day2.rs
use anyhow::Result;
use std::fs;

static DRAW: usize = 3;
static WIN: usize = 6;

fn to_number(token: &str) -> usize {
    match token {
        "A" | "X" => 1,
        "B" | "Y" => 2,
        "C" | "Z" => 3,
        _ => 0,
    }
}

fn play1(opponent: usize, player: usize) -> usize {
    return if opponent == player {
        DRAW + player
    } else if (opponent + 1) == player || opponent == 3 && player == 1 {
        WIN + player
    } else {
        player
    };
}

fn play2(opponent: usize, action: usize) -> usize {
    return if action == 2 {
        DRAW + opponent
    } else if action == 1 {
        if opponent == 1 {
            3
        } else {
            opponent - 1
        }
    } else {
        if opponent == 3 {
            1 + WIN
        } else {
            opponent + 1 + WIN
        }
    };
}

fn play_round(play: fn(usize, usize) -> usize) -> Result<usize> {
    return Ok(fs::read_to_string("../inputs/input2")?
        .split("\n")
        .filter_map(|line| {
            if line.is_empty() {
                return None;
            }
            let tokens: Vec<usize> = line
                .split(" ")
                .take(2)
                .map(|token| to_number(token))
                .collect();
            Some(tokens)
        })
        .map(|values| play(values[0], values[1]))
        .sum());
}

fn main() -> Result<()> {
    let part1: usize = play_round(play1)?;
    println!("Part 1: {part1}");
    let part2: usize = play_round(play2)?;
    println!("Part 2: {part2}");
    Ok(())
}
